### ISSUE TITLE
Lambda Cube

### ISSUE THREAD

**Issue by qdwang** [qdwang/notes#5]
#### Simply Typed Lambda Calculus: A Term Depends on a Term

In Simply Typed Lambda Calculus, Given a term M, we may form a new term
```
    λ x → M
```
which expects a term as argument
in other words, the term `λ x → M` depends on a term 
(given different arguments, the values of the term `λ x → M` are different)

#### Term depends on Type → λ2 (Polymorphism)
```
   | Agda Version Here!            | Haskell Version Here...      |
   | identity : {A : Set} → A → A  | identity : forall a . a -> a | ⇐ a is implicit
   | identity = λ x → x            | identity = \x -> x           |

   -- This is a explict type definition of identity
   identity : (A : Set) → A → A
   identity = λ A    →    λ x    →    x
                ↑           ↑
     (input a type) (type of x depend on A)       
```
   
#### Type depends on Type → λω (Type Constructor | Parametric Type)
```agda
    data List (A : Set) : Set where
      []   :              List A
      _::_ : A → List A → List A
```
`List` is a function which takes a type (like ℕ or 𝔹) and returns a new type (like [ℕ] or [𝔹]). 
In other words, it's a type which depends on another type.

#### Type depends on Term → λP (Indexed Type)
```agda
    zero? : ℕ → Set
    zero? 0 = ⊤
    zero? _ = ⊥
```
Different terms cause different type.

---
```agda
    data _≡ℕ_ : ℕ → ℕ → Set where
      z≡z :                    zero  ≡ℕ zero
      s≡s : ∀ {m n} → m ≡ℕ n → suc m ≡ℕ suc n
```
This is the equality on natural numbers, it has two cases:
1) `0 = 0`
2) given `m, n ∈ ℕ`, if `m = n` then `m+1 = n+1`

---
```agda
    data _≡𝔹_ : 𝔹 → 𝔹 → Set where
      t≡t :            true  ≡𝔹 true
      f≡f :            false ≡𝔹 false
       ↑                     ↑
(type constructor)     (cause different types)
```

#### λ2 + λω + λP = Dependent Types!
